peer/peerlist/list.go

```diff
@@ -0,0 +1,553 @@
+	if err := pl.once.WaitUntilRunning(ctx); err != nil {
+		return nil, nil, pl.newNotRunningError(err)
+	}
+
+	for {
+		if nextPeer := pl.choose(ctx, req); nextPeer != nil {
+			pl.notifyPeerAvailable()
+			nextPeer.StartRequest()
+			return nextPeer, pl.getOnFinishFunc(nextPeer), nil
+		}
+
+		if err := pl.waitForPeerAddedEvent(ctx); err != nil {
+			return nil, nil, err
+		}
+	}
+}
+
+func (pl *List) newNotRunningError(err error) error {
+	return yarpcerrors.FailedPreconditionErrorf("%s peer list is not running: %s", pl.name, err.Error())
+}
+
+// IsRunning returns whether the peer list is running.
+func (pl *List) IsRunning() bool {
+	return pl.once.IsRunning()
+}
+
+// choose grabs the next available peer from the PeerRing and returns it,
+// if there are no available peers it returns nil
+func (pl *List) choose(ctx context.Context, req *transport.Request) peer.Peer {
+	pl.lock.RLock()
+	defer pl.lock.RUnlock()
+	return pl.availableChooser.Choose(ctx, req)
+}
+
+// notifyPeerAvailable writes to a channel indicating that a Peer is currently
+// available for requests
+func (pl *List) notifyPeerAvailable() {
+	select {
+	case pl.peerAvailableEvent <- struct{}{}:
+	default:
+	}
+}
+
+// getOnFinishFunc creates a closure that will be run at the end of the request
+func (pl *List) getOnFinishFunc(p peer.Peer) func(error) {
+	return func(_ error) {
+		p.EndRequest()
+	}
+}
+
+// waitForPeerAddedEvent waits until a peer is added to the peer list or the
+// given context finishes.
+// Must NOT be run in a mutex.Lock()
+func (pl *List) waitForPeerAddedEvent(ctx context.Context) error {
+	if _, ok := ctx.Deadline(); !ok {
+		return pl.newNoContextDeadlineError()
+	}
+
+	select {
+	case <-pl.peerAvailableEvent:
+		return nil
+	case <-ctx.Done():
+		return pl.newUnavailableError(ctx.Err())
+	}
+}
+
+func (pl *List) newNoContextDeadlineError() error {
+	return yarpcerrors.Newf(yarpcerrors.CodeInvalidArgument, _noContextDeadlineError, pl.name)
+}
+
+func (pl *List) newUnavailableError(err error) error {
+	return yarpcerrors.UnavailableErrorf("%s peer list timed out waiting for peer: %s", pl.name, err.Error())
+}
+
+// NotifyStatusChanged receives status change notifications for peers in the
+// list.
+func (pl *List) NotifyStatusChanged(pid peer.Identifier) {
+	pl.lock.Lock()
+	t := pl.getThunk(pid)
+	pl.lock.Unlock()
+
+	if t != nil {
+		t.NotifyStatusChanged(pid)
+	}
+}
+
+// getThunk returns either the available or unavailable peer thunk.
+// Must be called under a lock.
+func (pl *List) getThunk(pid peer.Identifier) *peerThunk {
+	if t := pl.availablePeers[pid.Identifier()]; t != nil {
+		return t
+	}
+
+	if t := pl.unavailablePeers[pid.Identifier()]; t != nil {
+		return t
+	}
+
+	return nil
+}
+
+// notifyStatusChanged gets called by peer thunks
+func (pl *List) notifyStatusChanged(pid peer.Identifier) {
+	pl.lock.Lock()
+	defer pl.lock.Unlock()
+
+	if t := pl.availablePeers[pid.Identifier()]; t != nil {
```

Comment:
question: i noticed that peerheap implementation first removes unavailable and then adds: https://github.com/yarpc/yarpc-go/blob/cfb72130201ed7a9012a63f962e59d84581616f5/peer/x/peerheap/list.go#L125

is there a reason why PeerList is different?